File: kotlin/numbers/src/main/kotlin/numbers_iterative.kt
class NumbersIterative {
    fun fibonacci(n: Int): Int {
        return fibonacciIter(n, 0, 1)
    }

    private fun fibonacciIter(n: Int, acc2: Int, acc1: Int): Int {
        if(n <= 0) return acc2
        else if(n <= 2) return acc1 + acc2
        else return fibonacciIter(n - 1, acc1, acc1 + acc2)
    }

    fun factorial(n: Int): Int {
        return factorialIter(n, 1)
    }

    private fun factorialIter(n: Int, acc: Int): Int {
        if(n <= 1) return acc
        return factorialIter(n - 1, n * acc)
    }

    fun sumNumbers(n: Int): Int {
        return sumNumbersIter(n, 0)
    }

    private fun sumNumbersIter(n: Int, acc: Int): Int {
        if(n <= 0) return acc
        return sumNumbersIter(n - 1, n + acc)
    }
}

File: ruby/numbers/src/numbers_iterative.rb
module NumbersIterative
  def self.fibonacci(n)
    fibonacci_iter(n, 0, 1)
  end

  def self.fibonacci_iter(n, acc2, acc1)
    if n <= 0
      acc2
    elsif n == 1
      acc1
    else
      fibonacci_iter(n - 1, acc1, acc1 + acc2)
    end
  end

  def self.factorial(n)
    factorial_iter(n, 1)
  end

  def self.factorial_iter(n, acc)
    if n <= 1
      acc
    else
      factorial_iter(n - 1, n * acc)
    end
  end

  def self.sum_numbers(n)
    sum_numbers_iter(n, 0)
  end

  def self.sum_numbers_iter(n, acc)
    if n <= 0
      acc
    else
      sum_numbers_iter(n - 1, n + acc)
    end
  end
end